--- origsrc/emacs-25.1.50/lisp/net/tramp.el	2016-11-07 17:18:04.000000000 -0500
+++ src/emacs-25.1.50/lisp/net/tramp.el	2016-11-14 11:09:25.978621600 -0500
@@ -2007,7 +2007,8 @@ ARGS are the arguments OPERATION has bee
 		  'file-accessible-directory-p 'file-attributes
 		  'file-directory-p 'file-executable-p 'file-exists-p
 		  'file-local-copy 'file-modes
-		  'file-name-as-directory 'file-name-directory
+		  'file-name-as-directory 'file-name-case-insensitive-p
+                 'file-name-directory
 		  'file-name-nondirectory 'file-name-sans-versions
 		  'file-ownership-preserved-p 'file-readable-p
 		  'file-regular-p 'file-remote-p 'file-symlink-p 'file-truename
