diff --git a/src/process.c b/src/process.c
index 14d7b6d..9b096f9 100644
--- a/src/process.c
+++ b/src/process.c
@@ -5153,7 +5153,7 @@ wait_reading_process_output (intmax_t time_limit, int nsecs, int read_kbd,
 
 	      p = XPROCESS (proc);
 
-#ifdef GNU_LINUX
+#if 1
 	      /* getsockopt(,,SO_ERROR,,) is said to hang on some systems.
 		 So only use it on systems where it is known to work.  */
 	      {
