--- origsrc/emacs-25.0.50/src/sheap.c	2015-10-19 16:58:07.000000000 -0400
+++ src/emacs-25.0.50/src/sheap.c	2015-10-20 21:27:45.428447200 -0400
@@ -28,7 +28,7 @@ along with GNU Emacs.  If not, see <http
 #ifdef ENABLE_CHECKING
 #define STATIC_HEAP_SIZE	(28 * 1024 * 1024)
 #else
-#define STATIC_HEAP_SIZE	(19 * 1024 * 1024)
+#define STATIC_HEAP_SIZE	(11 * 1024 * 1024)
 #endif
 
 int debug_sheap = 0;
