=== modified file 'src/conf_post.h'
--- src/conf_post.h	2014-08-25 17:44:27 +0000
+++ src/conf_post.h	2014-08-30 19:47:56 +0000
@@ -179,6 +179,10 @@
 #if defined CYGWIN && defined HAVE_NTGUI
 # define NTGUI_UNICODE /* Cygwin runs only on UNICODE-supporting systems */
 # define _WIN32_WINNT 0x500 /* Win2k */
+/* The following was in /usr/include/string.h prior to Cygwin 1.7.33.  */
+#ifndef strnicmp
+#define strnicmp strncasecmp
+#endif
 #endif
 
 #ifdef emacs /* Don't do this for lib-src.  */

