=== modified file 'src/emacs.c'
--- src/emacs.c	2013-02-25 05:55:37 +0000
+++ src/emacs.c	2013-03-13 19:28:50 +0000
@@ -44,7 +44,7 @@
 #include "w32common.h"
 #endif
 
-#if defined HAVE_NTGUI && defined CYGWIN
+#if defined CYGWIN
 #include "cygw32.h"
 #endif
 
@@ -1348,7 +1348,7 @@
 #ifdef WINDOWSNT
       syms_of_ntproc ();
 #endif /* WINDOWSNT */
-#if defined CYGWIN && defined HAVE_NTGUI
+#if defined CYGWIN
       syms_of_cygw32 ();
 #endif
       syms_of_window ();

