=== modified file 'src/w32term.c'
--- src/w32term.c	2013-03-23 19:40:43 +0000
+++ src/w32term.c	2013-03-24 00:16:11 +0000
@@ -6662,7 +6662,7 @@
   Fset_input_mode (Qnil, Qnil, make_number (2), Qnil);
 
   {
-    DWORD input_locale_id = (DWORD) GetKeyboardLayout (0);
+    DWORD input_locale_id = ((DWORD_PTR) GetKeyboardLayout (0) & 0xffffffff);
     w32_keyboard_codepage =
       codepage_for_locale ((LCID) (input_locale_id & 0xffff));
   }

